### PR TITLE
remove limited gql backend

### PR DESCRIPTION
**Pull request by unclejustin**
Related to https://github.com/sourcegraph/sourcegraph/issues/41145

Alright, what's the big idea?

### The big idea

This PR is "simply" removing `CodeInsightsGqlBackendLimited`. To do that we:

1. Create a new Zustand store to hold our license info
2. Move license check into a new hook that saves results to the store
3. Update `useApi` to always return the full `CodeInsightsGqlBackend`
4. Update `useUiFeatures` to reference the store
5. Find all references of `CodeInsightsGqlBackendContext.UiFeatures` and update them to use `useUiFeatures`
6. Delete `useUiFeatures` tests. (see explanation below)
7. Add integration tests
8. Update jest tests
9. Update storybook stories

### Ok, and this helps us....how exactly?

By itself, this PR doesn't do much (other than add some much-needed test coverage). But it isolates the primary logic change from the next PR that will remove the Context and Provider. That will involve a lot of boilerplate that would muddy up the logic changes here. Removing the context gives us our big win. It will make refactoring the API **much** easier and makes navigating the code actually possible. i.e. cmd+click a method definition takes you to the definition, not the context placeholder.

### Footnotes

6. I deleted the test because right now, it only asserts `licensed` correctly references the appropriate backend. Since we only have one backend that check no longer makes sense. The updated test would just assert that the store saves itself, which is meaningless. I considered updating the test to check the dynamic functions, but that would be out of scope for this PR. I also considered leaving the test in place but empty, but that feels even more wrong. I suggest covering `useUiFeatures` implicitly with integration tests that consume it.

> Screenshot notes: the before and after screenshots should be identical for their respective license/unlicensed modes

#### Licensed

PixelSnap 2022-10-05 at 09 52 30

#### Unlicensed

PixelSnap 2022-10-05 at 09 53 32

#### Licensed

PixelSnap 2022-10-05 at 09 43 21

#### Unlicensed

PixelSnap 2022-10-05 at 09 47 50

## Test plan

A note on real vs fake testing. To test this locally you can "fake" the license quickly by just updating the license check in `use-license` by setting `licensed` to `true` or `false`.

This works for testing most of the UI, but the locked insight cards are only displayed if the license is actually updated (because that check happens on the backend).

If you want to perform a "real" test, you need to [update your license accordingly](https://docs.sourcegraph.com/dev/how-to/generate_license_key_for_testing).

- Integration tests should pass
- Set `licensed: true` 
- The licensed instance should render the dashboard without limited-mode banners.
- Set `licensed: false`
- The unlicensed instance should render the dashboard with limited-mode banners.

## App preview:

- [Web](https://sg-web-insights-remove-limited-backend.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-hlmgojwooj.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by vovakulikov** [comment]
@unclejustin I haven't reviewed it but just think before I forget this again, do you think it would be useful to have an integration test for unlicensed mode? I don't think have anything about it in our integrations tests at the moment.

**Comment by unclejustin** [comment]
@vovakulikov excellent suggestion. I'll update the issue to expect a passing test before and after removing the code.

File: client/web/src/enterprise/insights/hooks/use-license.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                    }
+                `,
+            })
+            .then(result => {
+                const licensed = result.data.enterpriseLicenseHasFeature
+                useCodeInsights.setState({ licensed, insightsLimit: licensed ? null : 2 })
+                setFetched(true)
+            })
+            .catch(() => new Error('Something went wrong fetching the license.'))
+    }, [apolloClient])
```
Comment (unclejustin) [inline client/web/src/enterprise/insights/hooks/use-license.ts:31]: There is some debate about how we get this information. Vova wants to add the license info to `window.context`, which would make this entire hook unnecessary. Since there is not a consensus yet, I've left this how it currently gets license info: with an API call.

If we ever move to `window.context` I would set the license info directly on the store when that app boots, and remove this hook entirely.

File: client/web/src/integration/insights/dashboards/render-empty-dashboard.test.ts [hunk 2]
```diff
@@ -53,5 +54,43 @@ describe('Code insights empty dashboard', () => {
...
+        )) as string
+        const addInsightsButtonCardText: string = (await driver.page.evaluate(
+            button => button.textContent,
+            addInsightsButtonCard
+        )) as string
+
+        assert(/empty dashboard/i.test(dashboardSelectButtonText))
+        assert(/add insights/i.test(addInsightsButtonCardText))
+
+        const gaModal = await driver.page.$('[role="dialog"][aria-label="Code Insights Ga information"]')
```
Comment (unclejustin) [inline client/web/src/integration/insights/dashboards/render-empty-dashboard.test.ts:92]: Note 1:

This only really covers one scenario of "unlicensed". The one where they have not accepted the modal yet.

I think this is fine for now. It asserts that our flag is doing what is expected for a single case.

In the future, we should probably add more tests for each of the unlicensed parts in the app. i.e., the creation flow.

Note 2:

I chose to only run this test on an empty dashboard because the file is smaller and there is no benefit to running this test in both.

File: client/web/src/enterprise/insights/CodeInsightsRouter.tsx [hunk 3]
```diff
@@ -30,9 +31,10 @@ export interface CodeInsightsRouterProps extends TelemetryProps {
 }
 
 export const CodeInsightsRouter: React.FunctionComponent<React.PropsWithChildren<CodeInsightsRouterProps>> = props => {
+    const fetched = useLicense()
     const api = useApi()
```
Comment (vovakulikov) [inline client/web/src/enterprise/insights/CodeInsightsRouter.tsx:35]: It seems that we don't need this useApi hook anymore, no? Can we inline creation API instance to be more explicit here?

File: client/web/src/enterprise/insights/hooks/use-license.ts (hunk 1 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/hooks/use-license.ts:31]: Bright future we (probably) will never see

File: client/web/src/enterprise/insights/pages/dashboards/creation/components/InsightsDashboardCreationContent.tsx [hunk 4]
```diff
@@ -47,8 +42,7 @@ export interface InsightsDashboardCreationContentProps {
 export const InsightsDashboardCreationContent: React.FunctionComponent<InsightsDashboardCreationContentProps> = props => {
     const { initialValues, owners, onSubmit, children } = props
 
-    const { UIFeatures } = useContext(CodeInsightsBackendContext)
-    const { licensed } = UIFeatures
+    const { licensed } = useUiFeatures()
```
Comment (vovakulikov) [inline client/web/src/enterprise/insights/pages/dashboards/creation/components/InsightsDashboardCreationContent.tsx:45]: It seems like that with `useCodeInsights` and `useUiFeatures` hook we have more than one place where we can read `licensed` data. I would prefer to have just one place for this data, simply because it will be easy to remember which hook I should use

File: client/web/src/enterprise/insights/pages/landing/getting-started/components/code-insights-examples/code-insight-example-card/CodeInsightExampleCard.tsx [hunk 5]
```diff
@@ -7,14 +7,16 @@ import { TelemetryProps } from '@sourcegraph/shared/src/telemetry/telemetryServi
 import { Button, Link, LegendItem, LegendList, ParentSize } from '@sourcegraph/wildcard'
 
 import { useSeriesToggle } from '../../../../../../../../insights/utils/use-series-toggle'
+import { useCodeInsights } from '../../../../../../../../stores'
```
Comment (vovakulikov) [inline client/web/src/enterprise/insights/pages/landing/getting-started/components/code-insights-examples/code-insight-example-card/CodeInsightExampleCard.tsx:10]: I would say we shouldn't store any code insight specific (and non core specific page data) into the global stores folder. I think we should create our own top level directory inside of insights project instead of global web app level stores directory

File: client/web/src/integration/insights/dashboards/render-empty-dashboard.test.ts (hunk 2 above)
Comment (vovakulikov) [reply client/web/src/integration/insights/dashboards/render-empty-dashboard.test.ts:92]: Nice, thanks for this test, I agree that we have to write more tests for unlicensed mode

**Review by vovakulikov — COMMENTED** [review]
Thanks for this PR @unclejustin this is a really cool work. I left a few minor comments about the code but I haven't tested it locally yet. I will do it as soon as I start my day tomorrow.

File: client/web/src/enterprise/insights/pages/landing/getting-started/components/code-insights-examples/code-insight-example-card/CodeInsightExampleCard.tsx (hunk 5 above)
Comment (unclejustin) [reply client/web/src/enterprise/insights/pages/landing/getting-started/components/code-insights-examples/code-insight-example-card/CodeInsightExampleCard.tsx:10]: In our [zustand comments](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@037622c6460c9743af7cb8a18e0cfaa24a2190d7/-/blob/client/web/src/stores/index.ts?L4), we explicitly require that all stores be in this directory. The linter will fail the build if you put them anywhere else.

File: client/web/src/enterprise/insights/hooks/use-ui-features.ts [hunk 6]
```diff
@@ -42,8 +43,8 @@ export interface UseUiFeatures {
 }
 
 export function useUiFeatures(): UseUiFeatures {
-    const { UIFeatures, getActiveInsightsCount } = useContext(CodeInsightsBackendContext)
-    const { licensed, insightsLimit } = UIFeatures
+    const { getActiveInsightsCount } = useContext(CodeInsightsBackendContext)
+    const { licensed, insightsLimit } = useCodeInsights()
```
[the remark this answers is not in the document]
Comment (unclejustin) [reply client/web/src/enterprise/insights/hooks/use-ui-features.ts:47]: Totally agree. Note: I thought was following the established convention (see `useNotepad`) but upon further inspection I see we don't really have one. I do see two other stores that use `useSomeNameState` so I will use that.

File: client/web/src/enterprise/insights/CodeInsightsRouter.tsx (hunk 3 above)
Comment (unclejustin) [reply client/web/src/enterprise/insights/CodeInsightsRouter.tsx:35]: We can, but I recommend waiting until the next PR where I remove the context provider completely. I believe it makes the changes more traceable.

File: client/web/src/enterprise/insights/pages/dashboards/creation/components/InsightsDashboardCreationContent.tsx (hunk 4 above)
Comment (unclejustin) [reply client/web/src/enterprise/insights/pages/dashboards/creation/components/InsightsDashboardCreationContent.tsx:45]: I like and agree with this idea. However, I would prefer to isolate that change to a dedicated PR. This PR is already 400+ lines of code, and making this update has just enough moving parts (async call inside a store) that I think we should wait.

**Review by vovakulikov — APPROVED** [review]
